extensions/wikidata/src/org/openrefine/wikidata/commands/LoginCommand.java

```diff
@@ -23,61 +23,177 @@
+
+        if (isBlank(username) || isBlank(password)) {
+            for (Cookie cookie : cookies) {
+                String value = cookie.getValue();
+                switch (cookie.getName()) {
+                    case USERNAME:
+                        username = value;
+                        break;
+                    case PASSWORD:
+                        password = value;
+                        break;
+                    default:
+                        break;
+                }
+            }
+            if (isNotBlank(username) && isNotBlank(password)) {
+                // If the credentials are read from cookies, we must remember it.
+                remember = true;
+            }
         }
-        response.setCharacterEncoding("UTF-8");
-        response.setHeader("Content-Type", "application/json");
-
-        Writer w = response.getWriter();
-        JsonGenerator writer = ParsingUtilities.mapper.getFactory().createGenerator(w);
-
-        writer.writeStartObject();
-        writer.writeBooleanField("logged_in", manager.isLoggedIn());
-        writer.writeStringField("username", manager.getUsername());
-        writer.writeEndObject();
-        writer.flush();
-        writer.close();
-        w.flush();
-        w.close();
+
+        String consumerToken = request.getParameter(CONSUMER_TOKEN);
+        String consumerSecret = request.getParameter(CONSUMER_SECRET);
+        String accessToken = request.getParameter(ACCESS_TOKEN);
+        String accessSecret = request.getParameter(ACCESS_SECRET);
+
+        if (isBlank(consumerToken)|| isBlank(consumerSecret) || isBlank(accessToken) || isBlank(accessSecret)) {
+            for (Cookie cookie : cookies) {
+                String value = cookie.getValue();
+                switch (cookie.getName()) {
+                    case CONSUMER_TOKEN:
+                        consumerToken = value;
+                        break;
+                    case CONSUMER_SECRET:
+                        consumerSecret = value;
+                        break;
+                    case ACCESS_TOKEN:
+                        accessToken = value;
+                        break;
+                    case ACCESS_SECRET:
+                        accessSecret = value;
+                        break;
+                    default:
+                        break;
+                }
+            }
+            if (isNotBlank(consumerToken) && isNotBlank(consumerSecret) && isNotBlank(accessToken) && isNotBlank(accessSecret)) {
+                remember = true;
+            }
+        }
+
+        if (isNotBlank(username) && isNotBlank(password)) {
+            manager.login(username, password);
+            // Once logged in with new credentials,
+            // the old credentials in cookies should be cleared.
+            if (manager.getConnection() != null && remember) {
+                setCookie(response, USERNAME, username);
+                setCookie(response, PASSWORD, password);
```

Comment:
Instead of storing the password as cookie, could we store Wikidata's cookie as a cookie ourselves? That should let us restore the connection later on, without storing the user's password at all.

Reply:
Good idea! This is helpful for logging in with password, but for the owner-only consumer, we still need to store the credentials in cookies.